crates/ra_arena/src/map.rs

```diff
@@ -0,0 +1,70 @@
+//! A map from arena IDs to some other type. Space requirement is O(highest ID).
+
+use std::marker::PhantomData;
+
+use super::ArenaId;
+
+/// A map from arena IDs to some other type. Space requirement is O(highest ID).
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ArenaMap<ID, T> {
+    v: Vec<Option<T>>,
+    _ty: PhantomData<ID>,
+}
+
+impl<ID: ArenaId, T> ArenaMap<ID, T> {
+    pub fn insert(&mut self, id: ID, t: T) {
+        let idx = Self::to_idx(id);
+        if self.v.capacity() <= idx {
+            self.v.reserve(idx + 1 - self.v.capacity());
+        }
+        if self.v.len() <= idx {
+            while self.v.len() <= idx {
+                self.v.push(None);
+            }
+        }
```

Comment:
This can be `self.v.resize(idx, None)` I think

Reply:
Ah, it'll requre `T: Clon` :weary:

Reply:
Only if we require `T: Clone`, sadly. There's also `self.v.resize(idx, || None)`, but that method isn't stable yet...